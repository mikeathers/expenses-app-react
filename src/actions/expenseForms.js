import database from "../firebase/firebase";

// ADD_EXPENSE_FORM
export const addExpenseForm = (expenseForm) => {
  return {
    type: "ADD_EXPENSE_FORM",
    expenseForm
  };
};

export const startAddExpenseForm = (expenseFormData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { name = "", totalCost = 0, createdAt = 0 } = expenseFormData;
    const expenseForm = { name, totalCost, createdAt };
    return database.ref(`users/${uid}/expenseForms/`).push(expenseForm).then((ref) => {
      dispatch(addExpenseForm({
        id: ref.key,
        ...expenseForm
      }));
    });
  };
};

// EDIT_EXPENSE_FORM
export const editExpenseForm = (id, updates) => {
  return {
    type: "EDIT_EXPENSE_FORM",
    id, 
    updates
  };
};

// REMOVE_EXPENSE_FORM
export const removeExpenseForm = ({id} = {}) => {
  return {
    type: "REMOVE_EXPENSE_FORM",
    id
  };
};

// SET_EXPENSE_FORMS
export const setExpenseForms = (expenseForms) => ({
  type: "SET_EXPENSE_FORMS",
  expenseForms
});

export const startSetExpenseForms = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenseForms`).once("value").then((snapshot) => {
      const expenseForms = [];
      snapshot.forEach((childSnapshot) => {
        expenseForms.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenseForms(expenseForms));
    });
  };
};