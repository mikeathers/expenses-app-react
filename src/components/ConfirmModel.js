import React from "react";
import Modal from "react-modal";

const ConfirmModal = (props) => (
  <Modal
    isOpen={props.modalIsOpen}
    contentLabel="Confirm"
    className="modal"
  >
    <h3 className="modal__title">Are you sure you want to remove this {props.type}?</h3> 
    <button className="button button--modal" onClick={props.onCloseModal}>Cancel</button>
    <button className="button button--modal" onClick={props.onRemove}>Remove</button>
  </Modal>
);

export default ConfirmModal;