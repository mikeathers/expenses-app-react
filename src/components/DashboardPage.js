import React from 'react';
import ExpenseFormsHeader from "./ExpenseForms/ExpenseFormsHeader";
import ExpenseFormsList from "./ExpenseForms/ExpenseFormsList";

const DashboardPage = () => (
  <div>
    <ExpenseFormsHeader />
    <ExpenseFormsList />
  </div>
);

export default DashboardPage;
