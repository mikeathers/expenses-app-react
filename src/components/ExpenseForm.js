import React from "react";
import ToggleDisplay from "react-toggle-display";
import { SingleDatePicker } from "react-dates";
import moment from "moment";

import TravelExpenseForm from "./TravelExpenseForm";
import OtherExpenseForm from "./OtherExpenseForm";

class ExpenseForm extends React.Component {
  constructor() {
    super();
    this.state = ({
      travelSelected: true,
      otherSelected: false,
      createdAt: moment(),
      origin: "",
      destination: "",
      odometerStart: 0,
      odometerEnd: 0,
      totalMiles: 0,
      totalCost: 0,
      calendarFocused: false,
    });
  }
  onExpenseTypeChange = (e) => {
    if (e.target.value === "travel") {
      this.setState({
        travelSelected: !this.state.travelSelected,
        otherSelected: false
      })
    } else if (e.target.value === "other") {
      this.setState({
        travelSelected: false,
        otherSelected: !this.state.otherSelected
      });        
    }
  };

  // Doesnt allow the date to be deleted. Only updates the state if there is a value in the date picker. 
  // If the state doesnt get updated the value of the datepicker wont change meaning it cant be deleted.
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    };
  }
  // sets calendarFocused to true or false depending on whether the calendar is focused.
  // needed by react-dates component
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  onSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit({
      date: this.state.createdAt.valueOf(),
      description: "Travel",
      origin: this.state.origin,
      destination: this.state.destination,
      odometerStart: this.state.odometerStart,
      odometerEnd: this.state.odometerEnd,
      totalMiles: this.state.totalMiles,
      totalCost: this.state.totalCost
    });
  }

  onHandleData = (expenseData) => {
    this.setState({
      origin: expenseData.origin,
      destination: expenseData.destination,
      odometerStart: expenseData.odometerStart,
      odometerEnd: expenseData.odometerEnd,
      totalMiles: expenseData.totalMiles,
      totalCost: expenseData.totalCost
    });
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>

        <h2>New Expense</h2>
        <div className="input-group">
        <select onChange={this.onExpenseTypeChange} className="select">
        <option value="travel">Travel</option>
        <option value="other">Other</option>          
      </select>
        <SingleDatePicker
        date={this.state.createdAt}
        onDateChange={this.onDateChange}
        focused={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}   
        //displayFormat={() => moment().format("DD/MM/YYYY")}
      />
        </div>
          

          <div className="input-group" >
            
          </div>
          <ToggleDisplay show={this.state.travelSelected}>
            <TravelExpenseForm onHandleData={this.onHandleData} />
          </ToggleDisplay>

          <ToggleDisplay show={this.state.otherSelected}>
            <OtherExpenseForm />
          </ToggleDisplay>

          <div className="input-group">
            <button className="button">Save Expense</button>
          </div>
          

    </form>
    );
  };
};

export default ExpenseForm;