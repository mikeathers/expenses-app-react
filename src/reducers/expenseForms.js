const expenseFormReducerDefaultState = [];

export default (state = expenseFormReducerDefaultState, action) => {
  switch (action.type) {
    
    case "ADD_EXPENSE_FORM": 
      return [
        ...state,
        action.expenseForm
      ];

    case "EDIT_EXPENSE_FORM": 
      return state.map((expenseForm) => {
        if (expenseForm.id === action.id) {
          return {
            ...expenseForm,
            ...action.updates
          };
        } else {
          return expenseForm; 
        }
      });

    case "REMOVE_EXPENSE_FORM": 
      return state.filter(({ id }) => id !== action.id );

    case "SET_EXPENSE_FORMS": 
      return action.expenseForms;
    
    default: 
      return state;
  };

};