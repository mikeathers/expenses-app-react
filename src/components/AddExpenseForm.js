import React from "react";
import { connect } from "react-redux";
import moment from "moment";
import uuid from "uuid";

import { startAddExpenseForm } from "../actions/expenseForms";


class AddExpenseForm extends React.Component {
  constructor() {
    super();

    this.state ={ 
      formName: ""
    };
  }

 onTextChange = (e) => {
  this.setState({
    formName: e.target.value
  });  
 };

 onAddForm = () => {
  const now = moment();
  const expenseForm = {
    id: uuid(),
    name: this.state.formName,
    createdAt: now.valueOf(),
    totalCost: 0,
    note: ""
  };

   this.props.startAddExpenseForm(expenseForm)
 }; 

  render() {
    return (
      <div>      
        <input onChange={this.onTextChange} type="text" placeholder="Expense form name" />
        <button onClick={this.onAddForm}>Add</button>
      </div>
    );
  } 
  
};

const mapStateToProps = (state) => ({
  expenseForms: state.expenseForms
})

const mapDispatchToProps = (dispatch) => ({
  startAddExpenseForm: (expenseForm) => dispatch(startAddExpenseForm(expenseForm))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpenseForm);