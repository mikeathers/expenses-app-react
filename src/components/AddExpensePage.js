import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import ExpenseFormSummary from "./ExpenseFormSummary";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

class AddExpensePage extends React.Component {

  onSubmit = (expenseData) => {
    const expense = {
      expenseFormId: this.props.match.params.id,
      ...expenseData
    }
    this.props.startAddExpense(expense);
    this.props.history.push(`/expenseForm/${this.props.match.params.id}`);
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <ExpenseFormSummary name={this.props.expenseForm.name} totalCost={this.props.expenseForm.totalCost}/>
            <Link className="button" to={`/expenseForm/${this.props.expenseForm.id}`}>Back</Link>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />          
        </div>
      </div>
    );
  };
};


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

const mapStateToProps = (state, props) => ({
  expenseForm: state.expenseForms.find((form) => form.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);