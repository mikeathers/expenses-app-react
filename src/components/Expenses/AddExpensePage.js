import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import ExpenseFormSummary from "../ExpenseForms/ExpenseFormSummary";
import ExpenseForm from "../Expenses/ExpenseForm";
import { startAddExpense, startEditExpense } from "../../actions/expenses";
import { startEditExpenseForm } from "../../actions/expenseForms";
 
class AddExpensePage extends React.Component {

  onSubmit = (expenseData) => {
    const expenseForm = this.props.expenseForm;
    const expenseFormId = expenseForm.id;    
    const expense = {
      expenseFormId,
      ...expenseData
    }
    const totalCost = expenseForm.totalCost += expense.totalCost;
    const newExpenseForm = {
      ...expenseForm,
      totalCost
    };
    
    this.props.startAddExpense(expense);
    this.props.startEditExpenseForm(expenseFormId, newExpenseForm);
    this.props.history.push(`/expenseForm/${this.props.match.params.id}`);
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <ExpenseFormSummary name={this.props.expenseForm.name} totalCost={this.props.expenseForm.totalCost}/>
          </div>
        </div>
        <div className="content-container">
          <h2>New Expense</h2>
          <ExpenseForm editExpense={false} onSubmit={this.onSubmit} />          
        </div>
      </div>
    );
  };
};


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  startEditExpenseForm: (id, updates) => dispatch(startEditExpenseForm(id, updates))
});

const mapStateToProps = (state, props) => ({
  expenseForm: state.expenseForms.find((form) => form.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);