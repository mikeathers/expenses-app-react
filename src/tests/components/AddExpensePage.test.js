import React from "react";
import { shallow } from "enzyme";

import { AddExpensePage } from "../../components/AddExpensePage";
import expenseForms from "../fixtures/expenseForms";

test("should render AddExpensePage correctly", () => {
  const wrapper = shallow(<AddExpensePage expenseForm={expenseForms[0]} />);
  expect(wrapper).toMatchSnapshot();
});