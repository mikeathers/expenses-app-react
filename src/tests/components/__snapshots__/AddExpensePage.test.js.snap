// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <ExpenseFormSummary
        name="January 2017"
        totalCost={200}
      />
      <Link
        className="button"
        replace={false}
        to="/expenseForm/123"
      >
        Back
      </Link>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm />
  </div>
</div>
`;
